File: backend/agents/pantry_agent.py
"""PantryAgent: cleans and validates pantry items from the frontend."""


class PantryAgent:
    """Simple pantry cleaning and validation logic."""

    ALLOWED_UNITS = {"g", "ml", "unit"}

    def clean_pantry_items(self, items: list[dict]) -> list[dict]:
        """Remove empty rows, normalize values, and validate units/amounts."""
        cleaned_items: list[dict] = []

        for raw_item in items:
            name = str(raw_item.get("name", "")).strip().lower()
            amount_raw = raw_item.get("amount")
            unit = str(raw_item.get("unit", "")).strip().lower()

            # Skip truly empty rows so users can keep blank input lines.
            if name == "" and (amount_raw in (None, "")) and unit == "":
                continue

            if not name:
                raise ValueError("Pantry item name is required.")

            try:
                amount = float(amount_raw)
            except (TypeError, ValueError):
                raise ValueError(f"Invalid amount for '{name}'. Amount must be numeric.")

            if amount <= 0:
                raise ValueError(f"Invalid amount for '{name}'. Amount must be greater than zero.")

            if unit not in self.ALLOWED_UNITS:
                raise ValueError(
                    f"Invalid unit for '{name}'. Allowed units are: g, ml, unit."
                )

            cleaned_items.append({"name": name, "amount": amount, "unit": unit})

        return cleaned_items

File: supabase/functions/api/agents/pantryAgent.ts
export class PantryValidationError extends Error {}

const ALLOWED_UNITS = new Set(["g", "ml", "unit"]);

function normalizeName(value: unknown): string {
  return String(value ?? "")
    .trim()
    .toLowerCase()
    .replace(/\s+/g, " ");
}

function normalizeUnit(value: unknown): string {
  return String(value ?? "").trim().toLowerCase();
}

function isEmptyAmount(value: unknown): boolean {
  return value === null || value === undefined || value === "";
}

export interface PantryItem {
  name: string;
  amount: number;
  unit: "g" | "ml" | "unit";
}

export function cleanPantryItems(items: unknown): PantryItem[] {
  if (!Array.isArray(items)) {
    throw new PantryValidationError("Pantry items must be an array.");
  }

  const cleaned: PantryItem[] = [];

  for (let i = 0; i < items.length; i += 1) {
    const row = items[i] as Record<string, unknown>;
    const name = normalizeName(row?.name);
    const unit = normalizeUnit(row?.unit);
    const amountRaw = row?.amount;

    if (name === "" && isEmptyAmount(amountRaw) && unit === "") {
      continue;
    }

    if (!name) {
      throw new PantryValidationError(`Invalid pantry item at row ${i + 1}: name is required.`);
    }

    const amount = Number(amountRaw);
    if (!Number.isFinite(amount)) {
      throw new PantryValidationError(`Invalid amount for '${name}'. Amount must be numeric.`);
    }

    if (amount <= 0) {
      throw new PantryValidationError(`Invalid amount for '${name}'. Amount must be greater than 0.`);
    }

    if (!ALLOWED_UNITS.has(unit)) {
      throw new PantryValidationError(`Invalid unit for '${name}'. Allowed units are: g, ml, unit.`);
    }

    cleaned.push({ name, amount, unit: unit as PantryItem["unit"] });
  }

  return cleaned;
}
